core/src/main/scala/reconciler/Reconciler.scala

```diff
@@ -119,6 +119,88 @@ abstract class Reconciler[S <: CustomResource[_, _], T <: ObjectResource] {
+                     sourceListResourceDefinition: ResourceDefinition[ListResource[S]],
+                     ec: ExecutionContext): Future[akka.stream.scaladsl.Source[WatchEvent[S], NotUsed]] = {
+    context.list[ListResource[S]].map { l =>
+      val initialSource = Source(l.items.map(l => WatchEvent(EventType.MODIFIED, l)))
+      val watchedSource = context.watchAllContinuously[S](Some(l.resourceVersion))
+
+      initialSource.concat(watchedSource)
+    }
+  }
+
+  def watchTargetRaw(implicit context: RequestContext,
+                     targetFormat: Format[T],
+                     targetListFormat: Format[ListResource[T]],
+                     targetResourceDefinition: ResourceDefinition[T],
+                     targetListResourceDefinition: ResourceDefinition[ListResource[T]],
+                     ec: ExecutionContext): Future[akka.stream.scaladsl.Source[WatchEvent[T], NotUsed]] = {
+    context.list[ListResource[T]].map { l =>
+      val initialSource = Source(l.items.map(l => WatchEvent(EventType.MODIFIED, l)))
+      val watchedSource = context.watchAllContinuously[T](Some(l.resourceVersion))
+
+      initialSource.concat(watchedSource)
+    }
+  }
+
+  def merge(implicit context: RequestContext,
+            sourceFormat: Format[S],
+            sourceListFormat: Format[ListResource[S]],
+            sourceResourceDefinition: ResourceDefinition[S],
+            targetFormat: Format[T],
+            targetListFormat: Format[ListResource[T]],
+            targetResourceDefinition: ResourceDefinition[T],
+            ec: ExecutionContext): akka.stream.scaladsl.Source[List[Event], NotUsed] = {
+
+    val sourceRaw = Source.fromFutureSource(watchSourceRaw)
+    val targetRaw = Source.fromFutureSource(watchTargetRaw)
+
+    val `WatchEvent[S]` = TypeCase[WatchEvent[S]]
+    val `WatchEvent[T]` = TypeCase[WatchEvent[T]]
+
+    Source
+      .combine(sourceRaw, targetRaw)(Merge(_))
+      .scan(Cache(List.empty, Map.empty)) { (acc, x) =>
+        x match {
+          case `WatchEvent[S]`(we) =>
+            val key = CacheEntryKey(name = we._object.name, namespace = we._object.namespace, kind = we._object.kind)
+            val targets = acc.cache.get(key).fold(List.empty[T])(_.ts)
+            val cache = acc.cache + (key -> CacheEntry(Some(we._object), targets))
+
+            we._type match {
+              case EventType.ADDED | EventType.MODIFIED => Cache(List(Modified(we._object, targets)), cache)
+              case EventType.DELETED                    => Cache(List(Deleted(we._object, targets)), cache)
+            }
+
+          case `WatchEvent[T]`(we) =>
+            // this assumes there's only one owner and they're in the same namespace
+            val owner = we._object.metadata.ownerReferences.head
+            val sourceKey = CacheEntryKey(name = owner.name, namespace = we._object.namespace, kind = owner.kind)
+
+            acc.cache.get(sourceKey) match {
```

Comment:
This is an interesting one... take the scenario where a _Source_ gets deleted. In this case you will get a delete event for the source which _should_ cause this code to removed it from the cache. Subsequently you will get a delete event for the _Target_. We probably should not be adding an entry back into the cache as it will cause a memory as it will never be removed as the parent does not exist. 

In the same way if you add a _Source_ all _Target_ events should be come after that event. 

There is, however, an assumption that the events are in order which they should be the case though may not be due to the combine using a Merge... but we fix that at a later stage using a window sort if necessary.

Reply:
Using the initial state, the `source` in `CacheEntry` case class is no longer an `Option` so we're not adding such entries again.